File: www/javascript/plugindownloader.js
$(function() {
    $(".pluginbutton").on("click", function() {
        var elem = $(this);
        elem.prop('disabled', true);
        $("button").prop("disabled", true);
        pluginName = elem.data('plugin');
        action = elem.val();

        $.ajax({
            type: 'GET',
            url: '/plugin_control?action=' + action + '&name=' + pluginName,
            elem: elem,
            action: action,
            pluginName: pluginName,
            success: function(return_string) {
                data = JSON.parse(return_string);
                switch(this.elem.val()){
                    case "Add":
                        if (data['success'] == true) {
                            this.elem.val("Remove");
                            this.elem.parent().parent().find("input[value='Clear Config']").prop('disabled', false);
                            this.elem.parent().parent().find("input[value='Clear Database']").prop('disabled', false);
                            alert(this.pluginName + " Downloaded");
                            counter(+1);
                        } else {
                            alert(data["message"]);
                            if (data['code']){
                                this.elem.val("Remove");
                                this.elem.parent().parent().find("input[value='Reload']").show();
                                break;
                            }
                        }
                        this.elem.prop('disabled', false);
                        break;
                    case "Remove":
                        if (data['success'] == true) {
                            this.elem.val("Add");
                            this.elem.parent().parent().find("input[value='Reload']").hide();
                            this.elem.parent().parent().find("input[value='Clear Config']").prop('disabled', true);
                            this.elem.parent().parent().find("input[value='Clear Database']").prop('disabled', true);
                            alert(this.pluginName + " Removed");
                            counter(-1);
                        } else {
                            alert(data["message"]);
                        }
                        this.elem.prop('disabled', false);
                        break;
                    case "Reload":
                        if (data['success'] == true) {
                            this.elem.hide();
                            counter(+1);
                            alert(this.pluginName + " Reloaded");
                        } else {
                            alert(data["message"]);
                        }
                        this.elem.prop('disabled', false);
                        break;
                    case "Clear Config":
                        if (data['success'] == true) {
                            this.elem.prop("disabled", true);
                            counter(+1);
                            alert(this.pluginName + " Config Cleared");
                        } else {
                            alert(data["message"]);
                        }
                        break;
                    case "Clear Database":
                        if (data['success'] == true) {
                            this.elem.prop("disabled", true);
                            counter(+1);
                            alert(this.pluginName + " Database Cleared");
                        } else {
                            alert(data["message"]);
                        }
                        break;
                    case "Start":
                        if (data['success'] == true) {
                            this.elem.val("Stop");
                            alert(this.pluginName + " Started");
                            this.elem.parent().parent().find("input[value='Clear Config']").prop('disabled', true);
                            this.elem.parent().parent().find("input[value='Clear Database']").prop('disabled', true);
                        } else {
                            alert(data["message"]);
                        }
                        this.elem.prop('disabled', false);
                        break;
                    case "Stop":
                        if (data['success'] == true) {
                            this.elem.val("Start");
                            alert(this.pluginName + " Stopped");
                            this.elem.parent().parent().find("input[value='Clear Config']").prop('disabled', false);
                            this.elem.parent().parent().find("input[value='Clear Database']").prop('disabled', false);
                        } else {
                            alert(data["message"]);
                        }
                        this.elem.prop('disabled', false);
                        break;
                    default:
                        alert(this.elem.val() + " Not Programed Yet")
                        this.elem.prop('disabled', false);
                        break;
                }
                counter(0)
            },
        });
    });
});

function counter(value){
    count = parseInt($("#plugin_count").val());
    count += value;
    $("#plugin_count").val(count);
    if (count == 0){
        $("button").prop("disabled", true);
    }else{
        $("button").prop("disabled", false);
    }

}

File: www/plugin_downloader.py
'''System for downloading plugins'''
import cherrypy
from libs.html_template import HTMLTEMPLATE
import libs.html_parts as html_parts
from system.plugin_downloader import TackemSystemPluginDownloader


class PluginDownloader(HTMLTEMPLATE):
    '''Plugin Downloader'''

    @cherrypy.expose
    def index(self) -> str:
        '''Index Page'''
        if not self._tackem_system.auth.is_admin():
            raise cherrypy.HTTPError(status=401)
        return self._template(
            plugin_download_page(True),
            False,
            javascript="plugin_downloader/javascript"
        )

    @cherrypy.expose
    def javascript(self) -> str:
        '''Javascript File'''
        if not self._tackem_system.auth.is_admin():
            raise cherrypy.HTTPError(status=401)
        return str(open("www/javascript/plugindownloader.js", "r").read())



def plugin_download_page(full_system: bool = True) -> str:
    '''returns the web page for choosing plugins'''
    TackemSystemPluginDownloader().get_github_plugins()
    TackemSystemPluginDownloader().get_local_plugins()
    html = "<h2>GITHUB PLUGINS</h2>"
    plugin_count = 0
    panels_html = ""
    for plugin in TackemSystemPluginDownloader().github_plugins:
        title = plugin['plugin_type'] + " - " + plugin['plugin_name']
        loaded = False
        if plugin['downloaded']:
            loaded = TackemSystemPluginDownloader().is_plugin_loaded(
                plugin['plugin_type'],
                plugin['plugin_name']
            )
            if loaded:
                plugin_count += 1

        clear_config = html_parts.input_button_with_data(
            "Clear Config",
            class_name="pluginbutton",
            data={'plugin':plugin['name']},
            outer_div=False,
            enabled=not loaded,
            visible=(full_system and plugin['downloaded'])
        )
        clear_database = html_parts.input_button_with_data(
            "Clear Database",
            class_name="pluginbutton",
            data={'plugin':plugin['name']},
            outer_div=False,
            enabled=not loaded,
            visible=(full_system and plugin['downloaded'])
        )
        add_remove = html_parts.input_button_with_data(
            ("Delete" if plugin['downloaded'] else "Download"),
            class_name="pluginbutton",
            data={'plugin':plugin['name']},
            outer_div=False
        )

        stop_start_action = ""
        stop_start_enabled = True
        stop_start_visible = False

        if full_system:
            stop_start_visible = plugin['downloaded']
            stop_start_action = "Stop" if loaded else "Start"
        else:
            stop_start_action = "Reload"
            stop_start_visible = plugin['downloaded'] and not loaded

        start_stop = html_parts.input_button_with_data(
            stop_start_action,
            class_name="pluginbutton",
            data={'plugin':plugin['name']},
            outer_div=False,
            enabled=stop_start_enabled,
            visible=stop_start_visible
        )
        readme = TackemSystemPluginDownloader().get_readme_as_html(
            plugin['plugin_type'],
            plugin['plugin_name']
        )
        if readme is None:
            readme = plugin['description']
        panels_html += html_parts.plugin_panel(title, readme, clear_config,
                                               clear_database, start_stop, add_remove)
    html += panels_html
    html += "<h2>LOCAL PLUGINS</h2>"
    panels_html = ""
    for plugin in TackemSystemPluginDownloader().local_plugins:
        if not plugin['offical']:
            title = plugin['plugin_type'] + " - " + plugin['plugin_name']
            offical = ""
            if plugin['repo']:
                offical = "[repo]"
            panels_html += "<H4>" + title + "</H4>" + offical
    html += panels_html
    html += html_parts.hidden("plugin_count", str(plugin_count), True)
    if full_system:
        return html
    html += html_parts.form("/", html_parts.hidden_page_index(3), "Next", "")
    if plugin_count == 0:
        html += '<script>$("button").prop("disabled", true);</script>'

    return html
